### PR TITLE
Nightlies

### PR DESCRIPTION
**Pull request by edunham**
2 separate config files because the Android upload to the servo-rust bucket in the Rust account, and the everything-else upload to the servo-developer-preview bucket in the Servo account, require different credentials. @aneeshusa, could you remind me of the right way to salt the near-identical files which end up with different pillar data and different names?

Windows nightly requires: 

* Appropriate package command
* Either refactoring the DynamicServoYAMLFactory to handle how all windows commands need to be run in their own bash shells (correct solution) or writing a new custom Windows factory (fast but bad solution)

---
This change is [Reviewable](https://reviewable.io/reviews/servo/saltfs/409)

### CODE REVIEW

**Comment by aneeshusa** [comment]
Just woke up and haven't looked at this yet, but you must be psychic  @edunham - I was about to open an issue/PR for this last night! This should also fix servo/servo#10339.

File: buildbot/master/files/config/master.cfg [hunk 1]
```diff
@@ -85,23 +85,29 @@ c['schedulers'].append(schedulers.SingleBranchScheduler(
...
     ],
 ))
 c['schedulers'].append(schedulers.Nightly(
     name="Nightly",
     branch="master",
-    builderNames=["android-nightly"],
+    builderNames=[
+        "android-nightly",
+        "linux-nightly",
+        "mac-nightly"
```
Comment (aneeshusa) [inline buildbot/master/files/config/master.cfg]: nit: leave a comma at the end of this line to reduce future diff noise

**Comment by aneeshusa** [comment]
I would prefer to use environment variables + `logEnviron: false` to pass the AWS IAM Key Id and Secret Access Key. Using a config file implies it's readable by any process running as `servo`, and we don't have any user isolation yet (e.g. IIRC Homu also runs as `servo`). Using environment variables limits the scope to which these credentials are exposed.
In the future I'd even like to see us using https://docs.aws.amazon.com/STS/latest/APIReference/API_GetSessionToken.html to generate temporary session tokens so even if a token gets leaked it's NBD.

This can be done by modifying the `make_step` functions and adding a new heurisitic for `s3cmd`. By the way - any reason we are using `s3cmd` instead of the AWS CLI?

File: buildbot/master/files/config/steps.yml [hunk 2]
```diff
@@ -22,6 +22,12 @@ mac-rel-css:
   - bash ./etc/ci/lockfile_changed.sh
   - bash ./etc/ci/manifest_changed.sh
 
+mac-nightly:
+  - ./mach build --release
+  - ./mach package --release
+  - s3cmd -c ~/.s3cfg-servo put target/release/*.tar.gz s3://servo-developer-preview/nightly/`date +%Y`/
```
Comment (aneeshusa) [inline buildbot/master/files/config/steps.yml]: Two things:
- Prefer "$(command)" form to backticks - [style guide](https://github.com/servo/servo/blob/master/docs/STYLE_GUIDE.md).
- I believe that since make_step splits the command up into a list before passing it to Buildbot, Buildbot will exec the command directly, instead of through a shell, and the command substitution won't work. IMO it would be better to make a script for this in the servo repo in the `etc/ci` directory and call it from here.

File: buildbot/master/files/config/steps.yml [hunk 3]
```diff
@@ -22,6 +22,12 @@ mac-rel-css:
...
+  - rm -rf target/
```
Comment (aneeshusa) [inline buildbot/master/files/config/steps.yml]: Why did you add the `rm -rf` lines? Buildbot will clean out the working directory at the start of each build IIRC (or I think maybe the `git` step does this), so it should not be necessary to have as a standalone step.

**Comment by aneeshusa** [comment]
It may also be nice to set up a [`MailNotifier`](http://docs.buildbot.net/current/manual/cfg-statustargets.html#mailnotifier) so Buildbot sends emails when a nightly build fails, but that can be saved for a follow-up.

File: buildbot/master/files/config/steps.yml (hunk 2 above)
Comment (ConnorGBrewster) [inline buildbot/master/files/config/steps.yml]: Also the mac package script now generates a `.dmg` instead of a `.tar.gz` and it is put in the `target` directory (not `debug` or `release` directories)

**Comment by metajack** [comment]
Is there a reason to continue to upload android builds to a Rust controlled bucket? Why don't we just dump everything in the servo bucket?

**Comment by larsbergstrom** [comment]
servo-master1 builds get:
```
[ FAIL ] Buildbot master config check failed:

         Configuration Errors:

           Unknown builder 'linux-nightly' in scheduler 'Nightly'

           Unknown builder 'mac-nightly' in scheduler 'Nightly'

           Unknown builder 'linux-nightly' in scheduler 'force'

           Unknown builder 'mac-nightly' in scheduler 'force'
```

**Comment by larsbergstrom** [comment]
I think they need to be added here:
https://github.com/edunham/saltfs/blob/281a9e39d65248a0e681b81086aa45330d1892cc/buildbot/master/files/config/master.cfg#L171

**Comment by aneeshusa** [comment]
I would rather not rush this PR too much, since we have some time before release and we can always make a build manually for release. @edunham if you want I can make a PR to your branch to implement passing credentials as environment variables.

**Comment by larsbergstrom** [comment]
I'm not super excited about manual builds from user machines and would rather hold off on releasing binaries until we do it from machines that were at least originally installed from a clean state with only the dependencies we've documented & listed in salt. Maybe manual builds from those machines would make me feel less stressed, but publishing bits to the world from laptops the team has taken through security checkpoints in all sorts of interesting countries seems dubious :-)

File: nightly/files/s3cfg-rust [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+get_continue = False
+gpg_command = /usr/bin/gpg
+gpg_decrypt = %(gpg_command)s -d --verbose --no-use-agent --batch --yes
+--passphrase-fd %(passphrase_fd)s -o %(output_file)s %(input_file)s
+gpg_encrypt = %(gpg_command)s -c --verbose --no-use-agent --batch --yes
+--passphrase-fd %(passphrase_fd)s -o %(output_file)s %(input_file)s
+gpg_passphrase =
+guess_mime_type = True
+host_base = s3.amazonaws.com
+host_bucket = %(bucket)s.s3.amazonaws.com
```
Comment (larsbergstrom) [inline nightly/files/s3cfg-rust]: Is the `s` after `%(bucket)` supposed to be there?

File: nightly/files/s3cfg-servo [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+get_continue = False
+gpg_command = /usr/bin/gpg
+gpg_decrypt = %(gpg_command)s -d --verbose --no-use-agent --batch --yes
+--passphrase-fd %(passphrase_fd)s -o %(output_file)s %(input_file)s
+gpg_encrypt = %(gpg_command)s -c --verbose --no-use-agent --batch --yes
+--passphrase-fd %(passphrase_fd)s -o %(output_file)s %(input_file)s
+gpg_passphrase =
+guess_mime_type = True
+host_base = s3.amazonaws.com
+host_bucket = %(bucket)s.s3.amazonaws.com
```
Comment (larsbergstrom) [inline nightly/files/s3cfg-servo]: Same question about the `s` here.

File: nightly/files/s3cfg-servo [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+skip_existing = False
+socket_timeout = 300
+stats = False
+stop_on_error = False
+storage_class =
+urlencoding_mode = normal
+use_https = True
+use_mime_magic = True
+verbosity = WARNING
+website_endpoint = http://%(bucket)s.s3-website-%(location)s.amazonaws.com/
```
Comment (larsbergstrom) [inline nightly/files/s3cfg-servo]: And here :-)

File: nightly/files/s3cfg-rust [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+skip_existing = False
+socket_timeout = 300
+stats = False
+stop_on_error = False
+storage_class =
+urlencoding_mode = normal
+use_https = True
+use_mime_magic = True
+verbosity = WARNING
+website_endpoint = http://%(bucket)s.s3-website-%(location)s.amazonaws.com/
```
Comment (larsbergstrom) [inline nightly/files/s3cfg-rust]: Same question :-)

**Comment by aneeshusa** [comment]
This will need servo/servo#11943 at the least.

**Comment by aneeshusa** [comment]
Filed #413 and #414 as follow-ups. Closing in favor of #410 (and #411).
